main.go

```diff
@@ -16,33 +16,206 @@ limitations under the License.
+	reader := csv.NewReader(inputFile)
+	reader.TrimLeadingSpace = true
+	line, err := reader.Read()
 	if err != nil {
-		panic(err)
+		return nil, err
+	}
+	var zoneNames []string
+	for _, name := range line[1:] {
+		name = strings.TrimSpace(name)
+		zoneNames = append(zoneNames, name)
 	}
+	var dataArray []inputData
+	for {
+		var data inputData
+		var done bool
+		data, done, err = readOneRow(zoneNames, reader)
+		if done {
+			break
+		}
+		if err != nil {
+			klog.Infof("can't parse input data: %v, skip to next row\n", data.name)
+			continue
+		}
+		dataArray = append(dataArray, data)
+	}
+	return dataArray, err
+}
 
-	var sim modeling.TheoreticalSimulator
+// parse one row of input file to one instance of inputData
+func readOneRow(zoneNames []string, reader *csv.Reader) (inputData, bool, error) {
+	rowCells, err := reader.Read()
+	if err == io.EOF {
+		return inputData{}, true, nil
+	}
+	if err != nil {
+		return inputData{}, true, err
+	}
+	var rowData inputData
+	rowData.name = rowCells[0]
+	for index, data := range rowCells[1:] {
+		nodestr := strings.Fields(data)
+		// convert string to int. number of nodes in a zone
+		numNode, err := strconv.Atoi(nodestr[0])
+		if err != nil {
+			return rowData, false, err
+		}
+		// convert string to int. number of endpoints in a zone
+		numEndpoints, err := strconv.Atoi(nodestr[1])
+		if err != nil {
+			return rowData, false, err
+		}
+		rowData.zones = append(rowData.zones, modeling.Zone{
+			Nodes:     numNode,
+			Endpoints: numEndpoints,
+			Name:      zoneNames[index],
+		})
+	}
+	return rowData, false, nil
+}
 
-	// TODO: operations on metrics rather than printing only
-	model, err := modeling.NewModel(zones, alg, sim)
+// startSimulation processes simulation on input data, produces instances of
+// outputData structure and returns a slice of them
+func startSimulation(algName string, inputArray []inputData) ([]outputData, error) {
+	// create algrithm based on the algorithm name delivered by the flag
+	alg := modeling.NewAlgorithm(algName)
+	// create simulation model, currently do calculation based on probability
+	// rather than real simulation.
+	model, err := modeling.NewModel(alg, modeling.TheoreticalSimulator{})
 	if err != nil {
-		panic(err)
+		return nil, err
+	}
+	var outputArray []outputData
+	for _, rowData := range inputArray {
+		err := model.UpdateRegion(rowData.zones)
+		if err != nil {
+			return outputArray, fmt.Errorf("input name : %s, %v", rowData.name, err)
+		}
+		simRes, err := model.StartSimulation()
+		if err != nil {
+			return outputArray, fmt.Errorf("input name : %s, %v", rowData.name, err)
+		}
+		var result outputData
+		result.name = rowData.name
+		result.result = simRes
+		outputArray = append(outputArray, result)
 	}
 
-	err = model.StartSimulation()
+	return outputArray, nil
+}
+
+// parseResult parses outputData to evaluation metrics and writes back to a
+// result file
+func parseResult(file string, outputArray []outputData) (err error) {
```

Comment:
There's a lot in `main.go`, ideally this file should be pretty minimal and call out to other packages. Not needed in this PR, but as we start to think about splitting this into multiple packages, I think we'll also want to move a lot of the content here out into a new package.

Reply:
Agreed, I was about to split these functions to other packages, but I think we might do the package split with the `modeling` as well. So I kept the main.go this way, going to create an issue for this TODO.